File: ruby/74.search-a-2d-matrix.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
def search_matrix(matrix, target)
   m = matrix.first.length
   n = matrix.length

   left = 0
   right = m * n

   while left < right
        mid = ((left + right) / 2).floor
        num = matrix[(mid/m).floor][mid % m]

        if target == num
            return true
        end

        if target < num
            right = mid
        else
            left += 1
        end
   end

   false
end

File: rust/src/74.search-a-2d-matrix.rs
impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        use std::cmp::Ordering;
        let m = matrix[0].len();
        let n = matrix.len();

        let mut left = 0;
        let mut right = m * n;

        while left < right {
            let mid = (left + right) / 2;
            let num = matrix[mid / m][mid % m];

            match num.cmp(&target) {
                Ordering::Equal => return true,
                Ordering::Less => left = mid + 1,
                Ordering::Greater => right = mid,
            }
        }
        false
    }
}
